### PR TITLE
Jax / PyTorch ViT models implementation

### PR DESCRIPTION
**Pull request by pomonam**
This PR implements Vision Transformer (ViT) on both JAX (adapted mainly from PR #80) and PyTorch.

- Renames the imagenet workload and reference submission directories to `imagenet_resnet` and adds directory `imagenet_vit`.
- Contains Jax ViT implmenetation from PR #80 .
- Contains PyTorch ViT implementation.
- Reuses a lot of code from iamgenet resnet workload. It also directory copies the imagenet resnet reference submission. The workload for ViT should be changed in the future.
- Passes reference submission tests and parameters number tests.
- PyTorch and Jax models output almost equivalent logits with the same initialization and inputs.

### CODE REVIEW

**Comment by fsschneider** [comment]
Hi @pomonam,

thanks for the awesome PR! Zack, Runa, and/or I will review it asap.

Could you move the `fastMRI` commits to a separate PR?
Could you also fix the linting issues?

**Comment by pomonam** [comment]
@fsschneider Thank you! 

Sorry about the incorrect branch commits. I reverted my mistakes (but in the process, accidentally created PR #84 and you can delete this PR).

This should fix all linting issues.

**Review by runame — APPROVED** [review]
LGTM now! I haven't compared the details of the Jax and PyTorch version of the ViT model, but since you compared the outputs given the same input, I think it's fine for now.

Did you already try to run the workload with the hyperparameter settings from the ImageNet ResNet workload? Finding a working hyperparameter config is probably the obvious next step.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/workload.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    torch.random.manual_seed(rng[0])
+    model_kwargs = models.decode_variant('S/16')
+    model = models.ViT(num_classes=1000, **model_kwargs)
+    self._param_shapes = {
+        k: spec.ShapeTuple(v.shape) for k, v in model.named_parameters()
+    }
+    model.to(DEVICE)
+    if N_GPUS > 1:
+      if PYTORCH_DDP:
+        model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
```
[the remark this answers is not in the document]
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/workload.py]: Thank you for the catch! I removed the line.

**Comment by pomonam** [comment]
> Did you already try to run the workload with the hyperparameter settings from the ImageNet ResNet workload? Finding a working hyperparameter config is probably the obvious next step.

No, I did not yet run the workload with the hyperparameter settings from the ImageNet ResNet workload. Finding a working hyperparameter config and the right target value is indeed the next step.

**Comment by runame** [comment]
> No, I did not yet run the workload with the hyperparameter settings from the ImageNet ResNet workload. Finding a working hyperparameter config and the right target value is indeed the next step.

Ok, cool, let me know if you need help with running things.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    out['pre_logits_2d'] = x_2d
+    out['pre_logits'] = x
+
+    if self.num_classes:
+      x_2d = out['logits_2d'] = self.head(x_2d)
+      x = out['logits'] = self.head(x)
+    return x, out
+
+
+def decode_variant(variant):
```
Comment (znado) [inline algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py:335]: given this is the same as the jax version, could we move this into a common file?

**Review by znado — APPROVED** [review]
LGTM! Thanks so much for this great code. It's awesome to see that the jax and pytorch models are close in outputs.

File: algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py (hunk 2 above)
Comment (pomonam) [reply algorithmic_efficiency/workloads/imagenet_vit/imagenet_pytorch/models.py:335]: Thank you! I moved this function into the shared `workload.py` file, which should be addressed in the commit [f6c542a](https://github.com/mlcommons/algorithmic-efficiency/pull/82/commits/f6c542a78d295d7a9c00dde8bd15fbde70f4ccc7).
